### ISSUE TITLE
`rake coverbands:baseline` dropped production database.

### ISSUE THREAD

**Issue by stephenprater** [danmayer/coverband#89]
First:  Let me be clear - this is not actually the gems fault.  The gem has a relatively minor security "flaw" which resulted in an outsize result in our environment because of poor practices / bad luck on _our_ end.

In the `baseline` task, Coverband loads (and because it's Ruby, executes) every `.rb` file under both `/lib` and `/app` - including potentially destructive scripts that are not loaded by the regular load process.

In our case, we had `lib/<submodule/db/schema.rb` files from an old submodule that were still present in Capistrano's cached copy.  Although we had run Coveralls on staging environments to verify it's operation, since that file was only present on a single production server, when we executed the `baseline` task on that server, it was loaded and forced schema creation, dropping existing tables.

Although this is definitely one of those "swiss cheese holes align" kind of problems - I think Coverband could benefit from being more intentional about what files it executes while establishing the baseline coverage.

Which of the following approaches would be most compatible with Coverband's overall philosophy?

1. Do not descend into submodule directories, and do not execute Ruby found there.
2. Maintain a 'blacklist' of "dangerous" files that should never be executed.
3. Open the database in "read only" mode while establishing a baseline.
4. Hook into the existing Rails eager load - which should eager load all files _referenced_ by the code base without loading potentially dangerous 'sidecar' scripts.

**Comment by latazzajones** [comment]
We've just been debugging a very similar issue - fortunately we caught ours in development. BUT! This seems like something that should be rectified or documented. In our case, we have an engine in our lib file with a dummy app and that schema is run by coverband. 

A blacklist seems like a good approach to this issue but this is also super dangerous behavior.

**Comment by danmayer** [comment]
WOW, prod DB drop that is terrible. So sorry about that... I am a bit confused what has changed to have two folks hit this recently, as I have never had a report like this in the 4 years the project has been in use.

I will need to ponder a bit... Some quick thoughts:

* I run baseline on dev machine and not on production, I just send it to the same source as when I report from production
    * If that isn't clear from the docs it might be good to make that more clear, there isn't any reason to run `baseline` on production (or staging, QA stacks, etc) EVER.
* I guess most ruby should 'execute' on load, but generally it shouldn't have any actions. I definitely consider any scripts that do pretty dangerous
* I don't like the blacklist as it would mean you need to know what to avoid before running. So I think we will need to come up with something better and safer than that.
   * That being said since both issues are schema related perhaps skipping all schema files by default would be a smart move.
* I do run Coverband on non Rails apps, so I would like to make sure the eager load approach could accommodate both by checking for availability and falling back to other behavior.

**Comment by danmayer** [comment]
I confirmed that `require './db/schema.rb'` on standard rails app would do this... since the schema is by default not in app or lib I haven't ever seen this or heard about it being an issue... make sense a bit more related to engines (although engines should still create folders and files in db folder). 

Thinking I will add `schema.rb` to default exclude list and merge the PR from @latazzajones as a first line of defense.

**Comment by danmayer** [comment]
ok @stephenprater coverband 1.5.2 was released with the PR from @latazzajones as well as this commit ( https://github.com/danmayer/coverband/commit/bf34b3b22dcf0c538d42982c3cf1eeaac388cd7f ) which will by default ignore any `schema.rb` files. 

I still think looking at your PR and other ways to avoid this ever happening to anyone else would be good... let me know thoughts on other steps. Also, so sorry again as I have had to be part of multiple production DB recoveries in the past and I know it is a terrible thing.
